sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -434,179 +518,122 @@ class GpuDynamicPartitionDataSingleWriter(
    * @param cachesMap used by `GpuDynamicPartitionDataConcurrentWriter` when fall back to single
    *                  writer, single writer should handle the stored writers and the pending caches
    */
-  def write(cb: ColumnarBatch,
+  protected def write(
+      cb: ColumnarBatch,
       cachesMap: Option[mutable.HashMap[String, WriterStatusWithCaches]]): Unit = {
     assert(isPartitioned)
     assert(!isBucketed)
 
-    // We have an entire batch that is sorted, so we need to split it up by key
-    var needToCloseBatch = true
-    var partitionColumns: Table = null
-    var distinctKeys: Table = null
-    var outputColumns: Table = null
-    var splits: Array[ContiguousTable] = null
-    var cbKeys: ColumnarBatch = null
     val maxRecordsPerFile = description.maxRecordsPerFile
+    val partDataTypes = description.partitionColumns.map(_.dataType).toArray
+    val outDataTypes = description.dataColumns.map(_.dataType).toArray
 
-    try {
-      partitionColumns = getPartitionColumns(cb)
-      val partDataTypes = description.partitionColumns.map(_.dataType).toArray
-      distinctKeys = distinctAndSort(partitionColumns)
-      val partitionIndexes = splitIndexes(partitionColumns, distinctKeys)
-      partitionColumns.close()
-      partitionColumns = null
-
-      // split the original data on the indexes
-      outputColumns = getOutputColumns(cb)
-      val outDataTypes = description.dataColumns.map(_.dataType).toArray
-      cb.close()
-      needToCloseBatch = false
-      splits = outputColumns.contiguousSplit(partitionIndexes: _*)
-      outputColumns.close()
-      outputColumns = null
-
-      cbKeys = copyToHostAsBatch(distinctKeys, partDataTypes)
-      distinctKeys.close()
-      distinctKeys = null
-
-      // Use the existing code to convert each row into a path. It would be nice to do this on
-      // the GPU, but the data should be small and there are things we cannot easily support
-      // on the GPU right now
-      import scala.collection.JavaConverters._
-      val paths = cbKeys.rowIterator().asScala.map(getPartitionPath)
-
-      paths.toArray.zip(splits).foreach(combined => {
-        val table = combined._2.getTable
-        val partPath = combined._1
-
+    // We have an entire batch that is sorted, so we need to split it up by key
+    // to get a batch per path
+    val (splits, paths) = splitBatchByKey(cb, partDataTypes)
+    withResource(splits) { _ =>
+      // NOTE: the `zip` here has the effect that will remove an extra `ContiguousTable`
+      // added at the end of `splits` because we use `upperBound` to find the split points,
+      // and the last split point is the number of rows.
+      splits.zip(paths).zipWithIndex.foreach { case ((partContigTable, partPath), partIx) =>
```

Comment:
Seems like part of the confusion here is that splitBatchByKey should be returning table, path pairs rather than two separate arrays that this goes out of its way to put back together.

Reply:
Yeah, I'll zip them while in `splitBatchByKey` and move the comment there...